### PR TITLE
feat(math): enable poseidon to hash packed prime fields

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR enables `Poseidon` and `Poseidon2` to hash packed prime fields.

### CODE REVIEW

File: tachyon/crypto/hashes/sponge/poseidon/poseidon_grain_lfsr.h [hunk 1]
```diff
@@ -101,8 +102,11 @@ PoseidonGrainLFSR<F>::PoseidonGrainLFSR(const PoseidonGrainLFSRConfig& config)
 }
 
 template <typename F>
-std::bitset<F::kModulusBits> PoseidonGrainLFSR<F>::GetBits(size_t num_bits) {
-  std::bitset<F::kModulusBits> ret;
+auto PoseidonGrainLFSR<F>::GetBits(size_t num_bits) {
```
Comment (Insun35) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon_grain_lfsr.h:105]: Do you have any reason to use `auto` here?

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_external_matrix_unittest.cc [hunk 2]
```diff
@@ -8,36 +8,40 @@
...
+                   Poseidon2Plonky3ExternalMatrix<math::PackedMersenne31>>;
 TYPED_TEST_SUITE(Poseidon2ExternalMatrixTest, MatrixTypes);
 
 TYPED_TEST(Poseidon2ExternalMatrixTest, DoApply) {
   using Matrix = TypeParam;
+  using F = typename Matrix::Field;
 
-  math::Vector<F> vector{
-      {{F::Random()}, {F::Random()}, {F::Random()}, {F::Random()}}};
+  math::Vector<F> vector{{{F(1)}, {F(1)}, {F(1)}, {F(1)}}};
```
Comment (Insun35) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_external_matrix_unittest.cc:36]:
```suggestion
  math::Vector<F> vector{{{F::One()}, {F::One()}, {F::One()}, {F::One()}}};
```

File: tachyon/crypto/hashes/sponge/poseidon/poseidon_grain_lfsr.h (hunk 1 above)
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon/poseidon_grain_lfsr.h:105]: Otherwise, `F::kModulusBits` should be changed to `std::conditional_t<math::FiniteFieldTraits<F>::kIsPackedPrimeField, typename math::FiniteFieldTraits<F>::PrimeField, F>::kModulusBits`!

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_external_matrix_unittest.cc (hunk 2 above)
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon2/poseidon2_external_matrix_unittest.cc:36]: I also agree that if we use F(1) for production, we should definitely use F::One(). However, since this is test code, it's beneficial to maintain this style for alignment purposes. For example:

   ```c++
   math::Vector<F> matrix{
     {F(0), F(1)},
     {F(2), F(3)},
   }
   ```

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/crypto/hashes/sponge/poseidon2/BUILD.bazel [hunk 3]
```diff
@@ -70,6 +71,7 @@ tachyon_cc_library(
     name = "poseidon2_plonky3_internal_matrix",
     hdrs = ["poseidon2_plonky3_internal_matrix.h"],
     deps = [
+        ":poseidon2_horizen_internal_matrix",
```
Comment (ashjeong) [inline tachyon/crypto/hashes/sponge/poseidon2/BUILD.bazel:74]: Does `"//tachyon/math/finite_fields:finite_field_traits"` not need to be added?

File: tachyon/crypto/hashes/sponge/poseidon/poseidon.h [hunk 4]
```diff
@@ -46,7 +46,42 @@ struct PoseidonSponge final : public PoseidonSpongeBase<PoseidonSponge<F>> {
     state.elements += config.ark.row(round_number);
   }
 
-  void ApplyMix(bool) { state.elements = config.mds * state.elements; }
+  void ApplyMix(bool) {
+    // NOTE (chokobole): Eigen matrix multiplication has a computational
+    // overhead compared to naive matrix multiplication.
```
Comment (ashjeong) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon.h:51]:
```suggestion
    // overhead unlike naive matrix multiplication.

File: benchmark/poseidon/Poseidon Benchmark MacM3.png
Comment (ashjeong) [inline benchmark/poseidon/Poseidon Benchmark MacM3.png:1]: Could you put the new benchmark results in a separate commit?

**Review by ashjeong — CHANGES_REQUESTED** [review]
https://github.com/kroma-network/tachyon/pull/435/commits/8a258acdccd167d05cbcf38183956abff1c3ba71
`NumBits()` and `Sum()` in `sponge.h` also use `<PrimeField>`! Please change as well

https://github.com/kroma-network/tachyon/pull/435/commits/bd71073fa78ee3cd7804133886a1bf76db12e584 I think it would be better to squash this with the previous commit

File: tachyon/math/finite_fields/baby_bear/packed_baby_bear.h [hunk 5]
```diff
@@ -25,6 +26,16 @@ using PackedBabyBear = PackedBabyBearAVX2;
 using PackedBabyBear = PackedBabyBearNeon;
 #endif
 
+template <>
+struct FiniteFieldTraits<PackedBabyBear> {
+  static constexpr bool kIsPrimeField = false;
+  static constexpr bool kIsPackedPrimeField = true;
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/baby_bear/packed_baby_bear.h:33]: Sounds good.

File: tachyon/math/finite_fields/packed_prime_field_unittest.cc [hunk 6]
```diff
@@ -196,4 +196,40 @@ TYPED_TEST(PackedPrimeFieldTest, Mul) {
   }
 }
 
+TYPED_TEST(PackedPrimeFieldTest, Div) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field_unittest.cc:201]: I see, then I'll change this test to check Inverse only! Since Inverse is correctly implemented in our structure, Division is also correctly implemented.

File: tachyon/math/finite_fields/packed_prime_field_unittest.cc [hunk 7]
```diff
@@ -196,4 +196,40 @@ TYPED_TEST(PackedPrimeFieldTest, Mul) {
...
+    PackedPrimeField a;
+    PackedPrimeField b;
+  } tests[] = {
+      {a, b},
+      {a, zero},
+      {a, one},
+  };
+
+  for (auto& test : tests) {
+    std::optional<PackedPrimeField> c = test.a / test.b;
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field_unittest.cc:219]: This part is unusual compared to other inversions or divisions. In BatchInverse(), if division by zero occurs, it is designed to populate the return value with zero.

File: tachyon/crypto/hashes/sponge/poseidon2/BUILD.bazel [hunk 8]
```diff
@@ -12,6 +12,7 @@ tachyon_cc_library(
         "//tachyon/base/buffer:copyable",
         "//tachyon/crypto/hashes/sponge:sponge_state",
         "//tachyon/crypto/hashes/sponge/poseidon:poseidon_sponge_base",
+        "//tachyon/math/finite_fields:finite_field_traits",
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon2/BUILD.bazel:15]: I'll add this to both of them, since this header is included in `poseidon2.h`, too.

**Review by ashjeong — APPROVED** [review]
LGTM~ Note that refactoring in regards to https://github.com/kroma-network/tachyon/issues/174 (as mentioned by @chokobole) should be done later!

File: tachyon/crypto/hashes/sponge/sponge_state.h [hunk 9]
```diff
@@ -3,29 +3,32 @@
...
   math::Vector<F> elements;
 
   // Current mode (whether its absorbing or squeezing)
   DuplexSpongeMode mode = DuplexSpongeMode::Absorbing();
 
-  PoseidonState() = default;
-  explicit PoseidonState(size_t size) : elements(size) {
+  SpongeState() = default;
+  explicit SpongeState(const SpongeConfig& config)
+      : SpongeState(config.rate + config.capacity) {}
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/sponge_state.h:30]: Why is this constructor added?

File: tachyon/math/finite_fields/packed_prime_field_base.h [hunk 10]
```diff
@@ -88,13 +88,13 @@ class PackedPrimeFieldBase : public Field<Derived> {
 
   // MultiplicativeGroup methods
   std::optional<Derived> Inverse() const {
-    NOTIMPLEMENTED();
     Derived ret;
+    CHECK(PrimeField::BatchInverse(values_, &ret.values_));
     return ret;
   }
 
   std::optional<Derived*> InverseInPlace() {
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field_base.h:96]:
```suggestion
  [[nodiscard]] std::optional<Derived*> InverseInPlace() {
```

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h [hunk 11]
```diff
@@ -58,10 +59,20 @@ struct Poseidon2Sponge final
...
+                             typename math::FiniteFieldTraits<F>::PrimeField,
+                             F>;
+
+      if constexpr (PrimeField::Config::kModulusBits <= 32) {
+        if (config.use_plonky3_internal_matrix) {
+          if constexpr (math::FiniteFieldTraits<F>::kIsPackedPrimeField) {
+            Poseidon2Plonky3InternalMatrix<F>::Apply(
+                state.elements, config.internal_diagonal_minus_one);
+          } else {
+            Poseidon2Plonky3InternalMatrix<F>::Apply(state.elements,
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h:73]: I think you should use PrimeField instead of F here.

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h [hunk 12]
```diff
@@ -62,12 +68,17 @@ struct Poseidon2Config : public PoseidonConfigBase<F> {
...
-      const std::array<F, N>& internal_diagonal_minus_one) {
+      const std::array<PrimeField, N>& internal_diagonal_minus_one) {
     Poseidon2ConfigEntry config_entry(rate, alpha, full_rounds, partial_rounds);
     Poseidon2Config ret = config_entry.ToPoseidon2Config<F>();
     ret.internal_diagonal_minus_one = math::Vector<F>(N);
     for (size_t i = 0; i < N; ++i) {
-      ret.internal_diagonal_minus_one[i] = internal_diagonal_minus_one[i];
+      if constexpr (math::FiniteFieldTraits<F>::kIsPackedPrimeField) {
+        ret.internal_diagonal_minus_one[i] =
+            F::Broadcast(internal_diagonal_minus_one[i]);
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h:78]: Use PrimeField instead of F.

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h [hunk 13]
```diff
@@ -79,9 +90,18 @@ struct Poseidon2Config : public PoseidonConfigBase<F> {
...
-    ret.internal_shifts = math::Vector<uint8_t>(N);
-    for (size_t i = 0; i < N; ++i) {
-      ret.internal_shifts[i] = internal_shifts[i];
+    ret.use_plonky3_internal_matrix = true;
+    if constexpr (math::FiniteFieldTraits<F>::kIsPackedPrimeField) {
+      ret.internal_diagonal_minus_one = math::Vector<F>(N + 1);
+      ret.internal_diagonal_minus_one[0] = F(PrimeField::Config::kModulus - 2);
+      for (size_t i = 1; i < N + 1; ++i) {
+        ret.internal_diagonal_minus_one[i] = F(1 << internal_shifts[i - 1]);
+      }
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h:99]: Use use PrimeField instead of F.

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h [hunk 14]
```diff
@@ -7,50 +7,74 @@
...
+      static_assert(PrimeField::Config::kModulusBits <= 32);
+      for (F& f : v) {
+        f *= F::Broadcast(PrimeField::FromMontgomery(1));
+      }
+    }
+  }
+
+  template <
+      typename F2 = F,
+      std::enable_if_t<math::FiniteFieldTraits<F2>::kIsPrimeField>* = nullptr>
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h:41]: Did you intend this function to be enabled for case where F2 is packed prime field as well?

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h [hunk 15]
```diff
@@ -7,50 +7,74 @@
...
         v[i] = FromU62(si);
       }
     }
   }
 
  private:
-  constexpr static PrimeField FromU62(uint64_t input) {
+  template <
+      typename F2 = F,
+      std::enable_if_t<math::FiniteFieldTraits<F2>::kIsPrimeField>* = nullptr>
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h:72]: Did you intend this function to be enabled for the case where F2 is packed prime field as well?

File: tachyon/crypto/hashes/sponge/poseidon/poseidon.h [hunk 16]
```diff
@@ -46,7 +46,42 @@ struct PoseidonSponge final : public PoseidonSpongeBase<PoseidonSponge<F>> {
     state.elements += config.ark.row(round_number);
   }
 
-  void ApplyMix(bool) { state.elements = config.mds * state.elements; }
+  void ApplyMix(bool) {
+    // NOTE (chokobole): Eigen matrix multiplication has a computational
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon.h:50]: Could you add a link to the Eigen matrix multiplication code please? Benchmarks would be good for sure but maybe code is enough.

File: tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h [hunk 17]
```diff
@@ -136,7 +136,10 @@ struct PoseidonSpongeBase : public FieldBasedCryptographicSponge<Derived> {
   bool Absorb(const T& input) {
     std::vector<F> elements;
     if (!SerializeToFieldElements(input, &elements)) return false;
+    return Absorb(absl::MakeConstSpan(elements));
+  }
 
+  bool Absorb(absl::Span<const F> input) {
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h:142]: I think this is not a good approach since the user has to know the fact that we have to explicitly call MakeConstSpan to avoid Serializing to Field elements. Can't we do this better with SFINAE?

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h (hunk 11 above)
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h:73]: I think it's fine because that must be `PrimeField` due to the above condition.

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h (hunk 12 above)
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h:78]: It's a `PackedPrimeField` here.

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h [hunk 18]
```diff
@@ -7,50 +7,76 @@
...
  public:
-  static void Apply(math::Vector<PrimeField>& v,
-                    const math::Vector<uint8_t>& shifts) {
+  template <typename F2 = F, std::enable_if_t<math::FiniteFieldTraits<
+                                 F2>::kIsPackedPrimeField>* = nullptr>
+  static void Apply(math::Vector<F>& v,
+                    const math::Vector<F>& diagonal_minus_one) {
+    using PrimeField =
+        std::conditional_t<math::FiniteFieldTraits<F>::kIsPackedPrimeField,
+                           typename math::FiniteFieldTraits<F>::PrimeField, F>;
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h:28]:
```suggestion
    using PrimeField = typename math::FiniteFieldTraits<F>::PrimeField;
```

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h [hunk 19]
```diff
@@ -7,50 +7,76 @@
...
+
+    Poseidon2HorizenInternalMatrix<F>::Apply(v, diagonal_minus_one);
+    if constexpr (PrimeField::Config::kUseMontgomery) {
+      static_assert(PrimeField::Config::kModulusBits <= 32);
+      for (F& f : v) {
+        f *= F::Broadcast(PrimeField::FromMontgomery(1));
+      }
+    }
+  }
```
Comment (TomTaehoonKim) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h:37]:
```suggestion
  template <typename F2 = F, std::enable_if_t<math::FiniteFieldTraits<
                                 F2>::kIsPackedPrimeField>* = nullptr>
  static void Apply(math::Vector<F>& v,
                    const math::Vector<F>& diagonal_minus_one) {
    Poseidon2HorizenInternalMatrix<F>::Apply(v, diagonal_minus_one);
    if constexpr (F::Config::kUseMontgomery) {
      static_assert(F::Config::kModulusBits <= 32);
      for (F& f : v) {
        f *= F::Broadcast(PrimeField::FromMontgomery(1));
      }
    }
  }
```

**Review by TomTaehoonKim — APPROVED** [review]
LGTM
